src/main/java/io/confluent/maven/resolver/ResolveKafkaVersionRangeMojo.java

```diff
@@ -0,0 +1,239 @@
+	private MavenProject project;
+
+	@Override
+	public void execute() throws MojoExecutionException, MojoFailureException {
+		VersionRangeRequest request = new VersionRangeRequest();
+		request.setRepositories(remoteRepositories);
+		DefaultArtifact artifact = new DefaultArtifact(groupId, artifactId, null, version);
+		request.setArtifact(artifact);
+
+		String constraintText = artifact.toString() + ", " + (includeSnapshots ? "including" : "excluding")
+				+ " snapshots";
+		getLog().info("Resolving range for " + constraintText);
+		try {
+			VersionRangeResult CEResult = repoSystem.resolveVersionRange(repoSession, request);
+			VersionRangeResult CCSResult = repoSystem.resolveVersionRange(repoSession, request);
+			Version highestCEVersion = fetchHighestKafkaVersion(CE, CEResult, constraintText);
+			Version highestCCSVersion = fetchHighestKafkaVersion(CCS, CCSResult, constraintText);
+
+			getLog().info("Highest " + CE + " version: " + highestCEVersion);
+			getLog().info("Highest " + CCS + " version: " + highestCCSVersion);
+
+			if (printCE) {
+				System.out.println(highestCEVersion);
+			}
+
+			if (printCCS) {
+				System.out.println(highestCCSVersion);
+			}
+
+			// Set CE property.
+			if (StringUtils.isNotEmpty(CEKAFKAVERSION) && project != null) {
+				getLog().info("Setting " + CEKAFKAVERSION + " property " + CEKAFKAVERSION + "=" + highestCEVersion);
+				project.getProperties().put(CEKAFKAVERSION, highestCEVersion.toString());
+			}
+
+			// Set CCS property.
+			if (StringUtils.isNotEmpty(CCSKAFKAVERSION) && project != null) {
+				getLog().info("Setting " + CCSKAFKAVERSION + " property " + CCSKAFKAVERSION + "=" + highestCCSVersion);
+				project.getProperties().put(CCSKAFKAVERSION, highestCCSVersion.toString());
+			}
+		} catch (VersionRangeResolutionException e) {
+			throw new MojoExecutionException("", e);
+		}
+	}
+
+	private Version fetchHighestKafkaVersion(String kafkaType, VersionRangeResult result, String constraintText)
+			throws MojoExecutionException, MojoFailureException {
+
+		// Workaround for https://issues.apache.org/jira/browse/MNG-3092
+		try {
+			getLog().debug(kafkaType + " version fetched " + result.getVersions());
+			if (!includeSnapshots) {
+				result.setVersions(removeSnapshots(result.getVersions()));
+			}
+			if (kafkaType.equals(CE)) {
+				result.setVersions(includeCE(result.getVersions()));
+				// if ce-kafka can not be fetched
+				if (result.getVersions().isEmpty()) {
+					getLog().info(CE + "can not be fetched");
+				}
+			} else if (kafkaType.equals(CCS)) {
+				result.setVersions(includeCCS(result.getVersions()));
+			} else {
+				throw new MojoFailureException("kafka type " + kafkaType + "wrong or not supported");
+			}
+
+			getLog().debug(kafkaType + " Constraint: " + result.getVersionConstraint());
+			getLog().info(kafkaType + " Versions in range: " + result.getVersions());
+
+			Version highestVersion = result.getHighestVersion();
```

Comment:
The `getHighestVersion` just return the top Version entry at the current Version list, actually the results at first from where I made request is already sorted, all I do essentially is just filter words.

Reply:
ok, i get it now. i thought the previous steps were figuring out the latest version but they are just getting a list of versions and this is where you get the latest one from the list. it would probably help to add some more comments around what is going on in this code, but i just didn't follow it for some reason. makes sense now